# <font color = dark_state_grey>Problem: Predict the flower class based on available attributes.</font>

In [1]:
#import datasets to load the pre-defined datasets
import numpy as np
import pandas as pd
import os
import seaborn as sb

In [2]:
#Set working directory
os.chdir("/Users/ajeet/Google Drive/DS/Projects/Classification")
os.getcwd()
df = pd.read_csv("iris.csv")

In [3]:
type(df)

pandas.core.frame.DataFrame

In [4]:
df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [5]:
df.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')

In [6]:
type(df['sepal_length'])

pandas.core.series.Series

In [7]:
type(df['sepal_width'])

pandas.core.series.Series

In [8]:
type(df['petal_length'])

pandas.core.series.Series

In [9]:
type(df['petal_width'])

pandas.core.series.Series

In [10]:
type(df['species'])

pandas.core.series.Series

We can tell pandas to automatically identify missing values if it knows our missing value marker.

In [11]:
df = pd.read_csv('iris.csv', na_values=['NA'])

In [12]:
type(df['species'])

pandas.core.series.Series

<font color = blue>Next, it's always a good idea to look at the distribution of our data — especially the outliers.</font>

In [13]:
df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [14]:
df['species'].unique()

array(['setosa', 'versicolor', 'virginica'], dtype=object)

In [15]:
df['species'].nunique()

3

## Missing value Analysis

In [16]:
#Check the missing values in whole data set
df.isnull().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

There are no missing values.

## Outlier Analysis

Box plot method to detect the outliers

In [17]:
import matplotlib.pyplot as plt

plt.boxplot(df['sepal_length'])

{'boxes': [<matplotlib.lines.Line2D at 0x10a659dd8>],
 'caps': [<matplotlib.lines.Line2D at 0x10a663828>,
 'fliers': [<matplotlib.lines.Line2D at 0x10a6744e0>],
 'means': [],
 'medians': [<matplotlib.lines.Line2D at 0x10a6740b8>],
 'whiskers': [<matplotlib.lines.Line2D at 0x10a659f28>,
  <matplotlib.lines.Line2D at 0x10a663400>]}

In [18]:
plt.boxplot(df['sepal_width'])

{'boxes': [<matplotlib.lines.Line2D at 0x10a674f98>],
 'caps': [<matplotlib.lines.Line2D at 0x10a680b70>,
 'fliers': [<matplotlib.lines.Line2D at 0x10a688828>],
 'means': [],
 'medians': [<matplotlib.lines.Line2D at 0x10a688400>],
 'whiskers': [<matplotlib.lines.Line2D at 0x10a680320>,
  <matplotlib.lines.Line2D at 0x10a680748>]}

<font color = red>There are total 4 outliers and total 150 rows</font> 

In [19]:
q75, q25 = np.percentile(df['sepal_width'], [75 ,25])
iqr = q75 - q25
 
min = q25 - (iqr*1.5)
max = q75 + (iqr*1.5)

In [20]:
min
max

4.05

In [21]:
df.shape

(150, 5)

In [22]:
df = df[df.sepal_width<max]

In [23]:
df = df[df.sepal_width>min]

In [24]:
plt.boxplot(df['sepal_width'])

{'boxes': [<matplotlib.lines.Line2D at 0x10a690550>],
 'caps': [<matplotlib.lines.Line2D at 0x10a690ef0>,
 'fliers': [<matplotlib.lines.Line2D at 0x10a698e80>],
 'means': [],
 'medians': [<matplotlib.lines.Line2D at 0x10a698a58>],
 'whiskers': [<matplotlib.lines.Line2D at 0x10a690978>,
  <matplotlib.lines.Line2D at 0x10a690da0>]}

In [25]:
df.shape

(146, 5)

In [26]:
plt.boxplot(df['petal_length'])

{'boxes': [<matplotlib.lines.Line2D at 0x10a6a1550>],
 'caps': [<matplotlib.lines.Line2D at 0x10a6a7320>,
 'fliers': [<matplotlib.lines.Line2D at 0x10a6a7f98>],
 'means': [],
 'medians': [<matplotlib.lines.Line2D at 0x10a6a7b70>],
 'whiskers': [<matplotlib.lines.Line2D at 0x10a6a1a90>,
  <matplotlib.lines.Line2D at 0x10a6a1eb8>]}

In [27]:
plt.boxplot(df['petal_width'])

{'boxes': [<matplotlib.lines.Line2D at 0x10a6b1898>],
 'caps': [<matplotlib.lines.Line2D at 0x10a6b8550>,
 'fliers': [<matplotlib.lines.Line2D at 0x10a6b8ef0>],
 'means': [],
 'medians': [<matplotlib.lines.Line2D at 0x10a6b8da0>],
 'whiskers': [<matplotlib.lines.Line2D at 0x10a6b1cc0>,
  <matplotlib.lines.Line2D at 0x10a6b1e10>]}

Now there no outlier present in the dataset

<font color = red>After removing the outliers now there are 146 obs is there</font> 

### Splitting the data into feature matrix and target variable vector

In [28]:
X = df.iloc[:, 0:4].values
y = df.iloc[:, 4].values

In [29]:
X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [4.9, 3.1, 1.5, 0.1],
       [5. , 3.2, 1.2, 0.2],
       [5.5, 3.5, 1.3, 0.2],
       [4.9, 3

In [30]:
df.tail()

sepal_length  sepal_width  petal_length  petal_width    species
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

In [31]:
y

array(['setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor'

<font color = red>We need to encode the dependent variable</font>

In [32]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

In [33]:
labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)

In [34]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

### Splitting the dataset into the Training set and Test set

In [35]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)


/Users/ajeet/anaconda3/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [36]:
lis = [X_train, X_test, y_train, y_test]
for i in range(4):
    print(len(lis[i]))

109
37
109
37


In [37]:
109+37

146

## Logistic Regression 

In [38]:
X_train

array([[5.4, 3.7, 1.5, 0.2],
       [7.7, 3. , 6.1, 2.3],
       [6.4, 2.7, 5.3, 1.9],
       [6.1, 2.9, 4.7, 1.4],
       [5.8, 2.8, 5.1, 2.4],
       [6.7, 3.1, 4.4, 1.4],
       [6.7, 3. , 5. , 1.7],
       [6.2, 2.8, 4.8, 1.8],
       [6.9, 3.1, 5.1, 2.3],
       [7.7, 2.6, 6.9, 2.3],
       [5.8, 2.7, 5.1, 1.9],
       [5.1, 2.5, 3. , 1.1],
       [5.6, 2.9, 3.6, 1.3],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 2.7, 3.9, 1.4],
       [5.5, 2.5, 4. , 1.3],
       [6.3, 2.8, 5.1, 1.5],
       [6.4, 2.8, 5.6, 2.1],
       [7.7, 2.8, 6.7, 2. ],
       [5.7, 2.9, 4.2, 1.3],
       [6.1, 3. , 4.9, 1.8],
       [6.4, 3.2, 4.5, 1.5],
       [6.7, 3.3, 5.7, 2.5],
       [6.5, 3. , 5.8, 2.2],
       [4.3, 3. , 1.1, 0.1],
       [5.9, 3.2, 4.8, 1.8],
       [4.9, 2.5, 4.5, 1.7],
       [5.5, 2.6, 4.4, 1.2],
       [7.4, 2.8, 6.1, 1.9],
       [5.4, 3.9, 1.3, 0.4],
       [5.7, 2.8, 4.5, 1.3],
       [4.6, 3.1, 1.5, 0.2],
       [6.7, 3.3, 5.7, 2.1],
       [6.4, 2.8, 5.6, 2.2],
       [4.6, 3

In [39]:
y_train

array([0, 2, 2, 1, 2, 1, 1, 2, 2, 2, 2, 1, 1, 0, 1, 1, 2, 2, 2, 1, 2, 1,
       2, 2, 0, 1, 2, 1, 2, 0, 1, 0, 2, 2, 0, 1, 2, 0, 1, 2, 1, 0, 0, 2,
       2, 0, 2, 0, 2, 2, 0, 0, 2, 0, 0, 0, 1, 2, 2, 0, 0, 0, 1, 1, 0, 0,
       1, 0, 2, 1, 2, 1, 0, 2, 0, 2, 0, 0, 1, 2, 2, 1, 1, 2, 1, 1, 2, 1,
       2, 0, 2, 1, 1, 0, 1, 2, 2, 0, 1, 1, 1, 1, 0, 0, 0, 2, 1, 2, 1])

In [40]:
#Fitting Logistic Regression to the Training set
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=0, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [41]:
X.shape

(146, 4)

### Predicting the Test set results

In [42]:
# Predicting the Test set results
y_pred = classifier.predict(X_test)

### Confusion Matrix

In [43]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

In [44]:
cm

array([[15,  0,  0],
       [ 0, 12,  1],
       [ 0,  0,  9]])

In [45]:
len(y_pred)

37

In [46]:
np.intersect1d(y_test, y_pred)

array([0, 1, 2])

In [47]:
y_test

array([0, 0, 0, 2, 2, 1, 0, 0, 2, 0, 0, 1, 2, 1, 1, 1, 0, 1, 2, 0, 1, 0,
       2, 1, 2, 0, 2, 1, 1, 2, 1, 0, 0, 1, 0, 0, 1])

In [48]:
Accuracy = (36/37)*100

In [49]:
Accuracy

97.2972972972973

In [50]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.972972972972973

In [51]:
sb.distplot(accuracy_score)
#plt.title('Average score: {}'.format(np.mean(accuracies)))

TypeError: unsupported operand type(s) for /: 'function' and 'int'

So, this might be the overfitting case

### Applying k-Fold Cross Validation

In [52]:
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 5)
accuracies.mean()


0.9270562770562771

In [53]:
accuracies

array([0.91666667, 0.90909091, 1.        , 0.85714286, 0.95238095])

### <font color = red>Accuracy = 92%</font>

In [54]:
sb.distplot(accuracies)
plt.title('Average score: {}'.format(np.mean(accuracies)))

Text(0.5,1,'Average score: 0.9270562770562771')

In [55]:
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from sklearn.svm import SVC
from sklearn.svm import LinearSVC
from sklearn.linear_model import SGDClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

In [56]:
models = []
models.append(("LoR", LogisticRegression()) )
models.append(("LDA", LinearDiscriminantAnalysis()) )
models.append(("QDA", QuadraticDiscriminantAnalysis()) )
models.append(("SVC", SVC()) )
models.append(("LSVC", LinearSVC()) )
models.append(("SGD", SGDClassifier()) )
models.append(("KNN", KNeighborsClassifier()) )
models.append(("GNB", GaussianNB() ))
models.append(("DT", DecisionTreeClassifier()) )
models.append(("RF", RandomForestClassifier()) )


In [57]:
for name, model in models:
     #cross validation among models, score based on accuracy
     accuracies = cross_val_score(estimator = model, X = X_train, y = y_train, cv = 5)
     print(name)
     print(accuracies.mean())
     print("\n")

LoR
0.9270562770562771


LDA
0.9635281385281387


QDA
0.9635281385281385


SVC
0.9642857142857142


LSVC
0.9444805194805195


SGD
0.5548701298701298


KNN
0.9453463203463205


GNB
0.9556277056277056


DT
0.9365800865800866


RF
0.9365800865800866




/Users/ajeet/anaconda3/lib/python3.6/site-packages/sklearn/discriminant_analysis.py:442: UserWarning: The priors do not sum to 1. Renormalizing
  UserWarning)
/Users/ajeet/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)
/Users/ajeet/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_ite